examples/decoding/plot_ssd_spatial_filters.py

```diff
@@ -28,144 +23,372 @@
+        self.filters_ = eigvects_[:, ix]
         self.patterns_ = np.linalg.pinv(self.filters_)
-
-    def transform(self, data):
-        if self.n_components is None:
-            n_components = len(self.picks_)
+        return self
+
+    def spectral_ratio_ssd(self, ssd_sources):
+        """
+        Spectral ratio measure for best n_components selection
+        See :footcite:'NikulinEtAl2011', Eq. (24).
+
+        Parameters
+        ----------
+        ssd_sources : array
+            data proyected on source space.
+
+        Returns
+        -------
+        spec_ratio : array, shape (n_channels)
+            array with the sprectal ratio value for each component
+        sorter_spec : array, shape (n_channels)
+            array of indeces for sorting spec_ratio.
+
+        """
+        psd, freqs = psd_array_welch(
+            ssd_sources, sfreq=self.sampling_freq, n_fft=self.n_fft)
+        sig_idx = _time_mask(freqs, *self.freqs_signal)
+        noise_idx = _time_mask(freqs, *self.freqs_noise)
+        if psd.ndim == 3:
+            mean_sig = psd[:, :, sig_idx].mean(axis=2).mean(axis=0)
+            mean_noise = psd[:, :, noise_idx].mean(axis=2).mean(axis=0)
+            spec_ratio = mean_sig / mean_noise
+        else:
+            mean_sig = psd[:, sig_idx].mean(axis=1)
+            mean_noise = psd[:, noise_idx].mean(axis=1)
+            spec_ratio = mean_sig / mean_noise
+        sorter_spec = spec_ratio.argsort()[::-1]
+        return spec_ratio, sorter_spec
+
+    def transform(self, inst):
+        """Estimate epochs sources given the SSD filters.
+
+        Parameters
+        ----------
+        inst : instance of Raw or Epochs (n_epochs, n_channels, n_times)
+            The data to be processed. The instance is modified inplace.
+
+        Returns
+        -------
+        X_ssd : instance of Raw or Epochs
+            The processed data.
+        """
+        if self.filters_ is None:
+            raise RuntimeError('No filters available. Please first call fit')
+        if self.return_filtered:
+            inst, _ = self.filter_data(inst)
+        if isinstance(inst, BaseRaw):
+            data = inst.get_data()
+            X_ssd = np.dot(self.filters_.T, data[self.picks_])
         else:
-            n_components = self.n_components
-        return np.dot(self.filters_.T[:n_components], data[self.picks_])
+            if not isinstance(inst, np.ndarray):
+                raise NotImplementedError()
+            self._check_X(inst)
+            data = inst
+            # project data on source space
+            X_ssd = np.asarray(
+                    [np.dot(self.filters_.T, epoch) for epoch in data])
+        if self.sort_by_spectral_ratio:
+            self.spec_ratio, self.sorter_spec = self.spectral_ratio_ssd(
+                ssd_sources=X_ssd)
+            self.filters_ = self.filters_[:, self.sorter_spec]
+            self.patterns_ = self.patterns_[self.sorter_spec]
+
```

Comment:
It depends essentially on whether sorted patterns are systematically shared across multiple related methods. In that case we should make sure that sorting is stable across repeated calls

Reply:
great I will check ! But I suppose that since there  the process is deterministic, nothing big should change between call of this method in the same conditions.  But essentially you are suggesting to avoid this within the API and just keep it as an atribute ?

Reply:
I think we can bookmark the issue for after merging. We should write unit tests for exactly this kind of doubt.